src/main/java/io/cryostat/core/agent/ProbeTemplate.java

```diff
@@ -0,0 +1,275 @@
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.core.agent;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.StringWriter;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.*;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+import javax.xml.transform.stream.StreamSource;
+
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+public class ProbeTemplate {
+
+    private static final String DEFAULT_FILE_NAME = "new_file.xml";
+    private static final String DEFAULT_CLASS_PREFIX = "__JFREvent";
+
+    private static final String XML_TAG_JFR_AGENT = "jfragent";
+    private static final String XML_TAG_CONFIG = "config";
+    private static final String XML_TAG_CLASS_PREFIX = "classprefix";
+    private static final String XML_TAG_ALLOW_TO_STRING = "allowtostring";
+    private static final String XML_TAG_ALLOW_CONVERTER = "allowconverter";
+    private static final String XML_TAG_EVENTS = "events";
+    private static final String XML_TAG_EVENT = "event";
+
+    private String fileName;
+    private String classPrefix;
+    private boolean allowToString;
+    private boolean allowConverter;
+
+    private List<Event> events;
+
+    public ProbeTemplate(
+            String fileName, String classPrefix, boolean allowToString, boolean allowConverter) {
+        this.fileName = fileName;
+        this.classPrefix = classPrefix;
+        this.allowToString = allowToString;
+        this.allowConverter = allowConverter;
+        this.events = new ArrayList<>();
+    }
+
+    public ProbeTemplate() {
+        this.fileName = DEFAULT_FILE_NAME;
+        this.classPrefix = DEFAULT_CLASS_PREFIX;
+        this.allowConverter = false;
+        this.allowToString = false;
+        this.events = new ArrayList<>();
+    }
+
+    public void deserialize(InputStream xmlStream) throws IOException, SAXException {
```

Comment:
What is the specific purpose of this `deserialize(InputStream)` method as opposed to the `deserialize(String)` defined later? This method appears to take an `InputStream` and consume it into a `byte[]` buffer, and then turn that buffer into a `String` and pass it to `deserialize(String)`. But, the first thing `deserialize(String)` does with its argument is to pull the `byte[]` back out of it and turn it into a `ByteArrayInputStream` to validate and then parse as XML.

Reply:
It seems to me like `deserialize(InputStream)` should be the real definition, but rather than consuming the stream into a buffer, it could just use some `BufferedInputStreams` to be able to validate and then "replay" the stream to parse. If there is some callsite that has the XML representation as a String in-memory already, then it can simply do `deserialize(new ByteArrayInputStream(xml.getBytes(StandardCharsets.UTF_8)))` as needed - or if that's a common enough use-case to include a convenience method here in the `ProbeTemplate` class, then it should just delegate to the `deserialize(InputStream)` version instead, the other way around.